crawler.py

```diff
@@ -326,6 +333,74 @@ def __crawl(self, current_url):
 
 			self.urls_to_crawl.add(link)
 
+	def write_sitemap_output(self):
+		are_multiple_sitemap_files_required = \
+			len(self.url_strings_to_output) > self.MAX_URLS_PER_SITEMAP
+
+		# When there are more than 50,000 URLs, the sitemap specification says we have
+		# to split the sitemap into multiple files using an index file that points to the
+		# location of each sitemap file.  For now, we require the caller to explicitly
+		# specify they want to create an index, even if there are more than 50,000 URLs,
+		# to maintain backward compatibility.
+		#
+		# See specification here:
+		# https://support.google.com/webmasters/answer/183668?hl=en
+		if are_multiple_sitemap_files_required and self.as_index:
+			self.write_index_and_sitemap_files()
+		else:
+			self.write_single_sitemap()
+
+	def write_single_sitemap(self):
+		self.write_sitemap_file(self.output_file, self.url_strings_to_output)
+
+	def write_index_and_sitemap_files(self):
+		sitemap_index_filename, sitemap_index_extension = os.path.splitext(self.output)
+
+		num_sitemap_files = math.ceil(len(self.url_strings_to_output) / self.MAX_URLS_PER_SITEMAP)
+		sitemap_filenames = []
+		for i in range(0, num_sitemap_files):
+			# name the individual sitemap files based on the name of the index file
+			sitemap_filename = sitemap_index_filename + '-' + str(i) + sitemap_index_extension
+			sitemap_filenames.append(sitemap_filename)
+
+		self.write_sitemap_index(sitemap_filenames)
+
+		for i, sitemap_filename in enumerate(sitemap_filenames):
+			self.write_subset_of_urls_to_sitemap(sitemap_filename, i * self.MAX_URLS_PER_SITEMAP)
+
+	def write_sitemap_index(self, sitemap_filenames):
+		sitemap_index_file = self.output_file
+		print(config.sitemapindex_header, file=sitemap_index_file)
+		for sitemap_filename in sitemap_filenames:
+			sitemap_url = urlunsplit([self.scheme, self.target_domain, sitemap_filename, '', ''])
+			print("<sitemap><loc>" + sitemap_url + "</loc>""</sitemap>", file=sitemap_index_file)
+		print(config.sitemapindex_footer, file=sitemap_index_file)
+
+	def write_subset_of_urls_to_sitemap(self, filename, index):
+		# Writes a maximum of self.MAX_URLS_PER_SITEMAP urls to a sitemap file
+		#
+		# filename: name of the file to write the sitemap to
+		# index:    zero-based index from which to start writing url strings contained in
+		#           self.url_strings_to_output
+		try:
+			sitemap_file = open(filename, 'w')
```

Comment:
Wonder if it would be good to use a context manager here, like `with open(filename, 'w') as sitemap_file:`?

Reply:
Agreed, definitely; changed.